README.md
# Maquiavelog
An easy to use log for Node.js.


CHANGELOG.md
# Maquiavelog Changelog

## Known issues

None.

## Unreleased

### Functionality added

- Set up Node.js project.
- Add log levels logic.
- Add outputs logic.

### Documentation

- Add Changelog file.
- Add Readme file.


test/unit/lib/levels.spec.js
/**
 * levels.spec.js
 * Unit tests for log levels.
 * 
 * Copyright (c) 2023 Ricardo Mendoza.
 * Licensed under the MIT License.
 */

'use strict';

const chai = require('chai');
const logLevels = require('../../../lib/levels');
const expect = chai.expect;

describe('Log levels', () => {
    let logFake;
    const testLevel = { label: 'info', value: 30 };
    const defaultLogLevels = {
        trace: 10,
        debug: 20,
        info: 30,
        warn: 40,
        error: 50,
        fatal: 60,
        silent: Infinity
    };

    beforeEach(() => {
        logFake = Object.assign({ _level: null }, logLevels);
    });

    it('has the default log levels set', () => {
        expect(logFake.levels).to.deep.equal(defaultLogLevels);
    });

    describe('Get level', () => {
        it('gets the current log level set if no parameters are provided', () => {
            logFake._level = testLevel;
            expect(logFake.getLevel()).to.deep.equal(testLevel);
        });

        it('gets the level when the level label is provided', () => {
            expect(logFake.getLevel(testLevel.label)).to.deep.equal(testLevel);
        });

        it('gets the level when the level value is provided', () => {
            expect(logFake.getLevel(testLevel.value)).to.deep.equal(testLevel);
        });

        it('gets no level if the level label provided is invalid', () => {
            expect(logFake.getLevel('invalid')).to.be.null;
        });

        it('gets no level if the level value provided is invalid', () => {
            expect(logFake.getLevel(0)).to.be.null;
        });

        it('gets no level if the type of the data provided is invalid', () => {
            expect(logFake.getLevel(true)).to.be.null;
        })
    });

    describe('Set level', () => {
        it('sets the log level when using the level label', () => {
            expect(logFake.getLevel()).to.be.null;
            logFake.setLevel(testLevel.label);
            expect(logFake.getLevel()).to.deep.equal(testLevel);
        });
    
        it('sets the log level when using the level value', () => {
            expect(logFake.getLevel()).to.be.null;
            logFake.setLevel(testLevel.value);
            expect(logFake.getLevel()).to.deep.equal(testLevel);
        });
    
        it('prevents the log level from being set when passing an invalid level label', () => {
            expect(logFake.setLevel.bind(logFake, 'invalid')).to.throw('The level provided is not valid');
        });
    
        it('prevents the log level from being set when passing an invalid level value', () => {
            expect(logFake.setLevel.bind(logFake, 1234)).to.throw('The level provided is not valid');
        });

        it('prevents the log level from being set when the type of the value passed is invalid', () => {
            expect(logFake.setLevel.bind(logFake, true)).to.throw('The level provided is not valid');
        });
    });

    describe('Is level enabled?', () => {
        it('checks if the level is enabled when tested with the level label', () => {
            expect(logFake.isLevelEnabled('trace')).to.equal(false);
            logFake.setLevel(defaultLogLevels.trace);
            expect(logFake.isLevelEnabled('trace')).to.equal(true);
        });
    
        it('checks if the level is enabled when tested with the level value', () => {
            expect(logFake.isLevelEnabled(10)).to.equal(false);
            logFake.setLevel(defaultLogLevels.trace);
            expect(logFake.isLevelEnabled(10)).to.equal(true);
        });
    
        it('treats invalid values as disabled levels when checking if a level is enabled', () => {
            logFake.setLevel(defaultLogLevels.trace);
            expect(logFake.isLevelEnabled('invalid')).to.equal(false);
            expect(logFake.isLevelEnabled(1234)).to.equal(false);
        });
    
        it('considers the level as enabled if it is equal to the level set', () => {
            logFake.setLevel(defaultLogLevels.warn);
            expect(logFake.isLevelEnabled('warn')).to.equal(true);
        });
    
        it('considers the level as enabled if it is greater than the level set and is not the silent level', () => {
            logFake.setLevel(defaultLogLevels.warn);
            expect(logFake.isLevelEnabled('error')).to.equal(true);
            expect(logFake.isLevelEnabled('silent')).to.equal(false);
        });
    
        it('considers the level as disabled if it is less than the level set', () => {
            logFake.setLevel(defaultLogLevels.warn);
            expect(logFake.isLevelEnabled('info')).to.equal(false);
        });
    
        it('considers all levels but silent as enabled if the trace level is set', () => {
            logFake.setLevel(defaultLogLevels.trace);
            expect(logFake.isLevelEnabled('trace')).to.equal(true);
            expect(logFake.isLevelEnabled('debug')).to.equal(true);
            expect(logFake.isLevelEnabled('info')).to.equal(true);
            expect(logFake.isLevelEnabled('warn')).to.equal(true);
            expect(logFake.isLevelEnabled('error')).to.equal(true);
            expect(logFake.isLevelEnabled('fatal')).to.equal(true);
            expect(logFake.isLevelEnabled('silent')).to.equal(false);
        });
    
        it('considers all levels as disabled if the silent level is set', () => {
            logFake.setLevel(defaultLogLevels.silent);
            expect(logFake.isLevelEnabled('trace')).to.equal(false);
            expect(logFake.isLevelEnabled('debug')).to.equal(false);
            expect(logFake.isLevelEnabled('info')).to.equal(false);
            expect(logFake.isLevelEnabled('warn')).to.equal(false);
            expect(logFake.isLevelEnabled('error')).to.equal(false);
            expect(logFake.isLevelEnabled('fatal')).to.equal(false);
            expect(logFake.isLevelEnabled('silent')).to.equal(true);
        });
    });
});


test/unit/lib/outputs.spec.js
/**
 * outputs.spec.js
 * Unit tests for outputs.
 * 
 * Copyright (c) 2023 Ricardo Mendoza.
 * Licensed under the MIT License.
 */

'use strict';

const chai = require('chai');
const Stream = require('stream');
const outputs = require('../../../lib/outputs');
const expect = chai.expect;

describe('Outputs', () => {
    let logFake;
    const testStream = new Stream.Writable();
    const testLevel = { label: 'info', value: 30 };

    beforeEach(() => {
        logFake = Object.assign({ _level: testLevel, _outputs: [] }, outputs);
    });

    describe('Add output', () => {
        it('adds an output', () => {
            expect(logFake._outputs).to.be.an('array').that.is.empty;
            const output = { stream: testStream, level: testLevel };
            logFake.addOutput(output);
            expect(logFake._outputs.length).to.equal(1);
        });
    
        it('adds an output when passing a level label to specify the level', () => {
            const output = { stream: testStream, level: testLevel.label };
            logFake.addOutput(output);
            expect(logFake._outputs[0].level).to.deep.equal(testLevel);
        });
    
        it('adds an output when passing a level value to specify the level', () => {
            const output = { stream: testStream, level: testLevel.value };
            logFake.addOutput(output);
            expect(logFake._outputs[0].level).to.deep.equal(testLevel);
        });
    
        it('adds an output with the current log level if no other level is specified', () => {
            const output = { stream: testStream };
            logFake.addOutput(output);
            expect(logFake._outputs[0].level).to.deep.equal(testLevel);
        });
    
        it('rejects the output if the specified level is invalid', () => {
            const output = { stream: testStream, level: 'invalid' };
            expect(logFake.addOutput.bind(logFake, output)).to.throw('The level provided is not valid');
        });
    
        it('rejects the output if it does not include a stream', () => {
            const output = { wrong: 'structure' };
            expect(logFake.addOutput.bind(logFake, output)).to.throw('No stream provided');
        });
    
        it('rejects the output if the specified stream does not have a write method', () => {
            const output = { stream: { write: 'string' } };
            expect(logFake.addOutput.bind(logFake, output)).to.throw('The stream provided does not have a write method');
        });
    
        it('ignores any extra properties from the provided output and sets only the stream and the level', () => {
            const output = { stream: testStream, level: testLevel, extra: 'property' };
            logFake.addOutput(output);
            expect(Object.keys(logFake._outputs[0]).length).to.equal(2);
            expect(logFake._outputs[0].stream).to.deep.equal(testStream);
            expect(logFake._outputs[0].level).to.deep.equal(testLevel);
        });
    });
});


test/unit/maquiavelog.spec.js
'use strict';

const Maquiavelog = require('../../maquiavelog');
const chai = require('chai');
const sinon = require('sinon');
const expect = chai.expect;

chai.use(require('sinon-chai'));

describe('Maquiavelog', () => {
    let maquiavelog;

    beforeEach(() =>{
        maquiavelog = Maquiavelog();
    });

    it('logs info("Hello World")', () => {
        sinon.stub(process.stdout, 'write').callsFake(chunk => {});
        maquiavelog.info('Hello world!');
        expect(process.stdout.write).to.have.been.calledOnceWith('Hello world!');
        sinon.restore();
    });
});


lib/levels.js
/**
 * levels.js
 * Main logic for log levels.
 * 
 * Copyright (c) 2023 Ricardo Mendoza.
 * Licensed under the MIT License.
 */

'use strict';

const _levelsByLabel = {
    trace: 10,
    debug: 20,
    info: 30,
    warn: 40,
    error: 50,
    fatal: 60,
    silent: Infinity
};

const _levelsByValue = {
    10: 'trace',
    20: 'debug',
    30: 'info',
    40: 'warn',
    50: 'error',
    60: 'fatal',
    Infinity: 'silent'
};

function getLevel(level) {
    if (level === undefined || level === null) {
        return this._level;
    }

    switch (typeof level) {
        case 'string':
            const label = level.toLowerCase();
            return _levelsByLabel[label] ? { label, value: _levelsByLabel[label] } : null;
        case 'number':
            return _levelsByValue[level] ? { label: _levelsByValue[level], value: level } : null;
        default:
            return null;
    }
}

function setLevel(level) {
    if (typeof level === 'string') {
        const label = level.toLowerCase();

        if (_levelsByLabel[label]) {
            this._level = { label, value: _levelsByLabel[label] };
            return;
        }
    } else if (typeof level === 'number' && _levelsByValue[level]) {
        this._level = { label: _levelsByValue[level], value: level };
        return;
    }

    throw new Error('The level provided is not valid');
}

function isLevelEnabled(level) {
    if (!this._level) {
        return false;
    }

    let label;

    if (typeof level === 'string') {
        label = level.toLowerCase();
    } else if (typeof level === 'number' && _levelsByValue[level]) {
        label = _levelsByValue[level];
    }

    if (label === 'silent') {
        return this._level.label === label;
    } else {
        return label && _levelsByLabel[label] ? _levelsByLabel[label] >= _levelsByLabel[this._level.label] : false;
    }
}

module.exports = {
    levels: _levelsByLabel,
    getLevel,
    setLevel,
    isLevelEnabled
};
